### PR TITLE
Enable outputFormatting when encoding to XML

### PR DESCRIPTION
**Pull request by alextov**

### CODE REVIEW

**Comment by jseibert** [comment]
Please provide before/after usage again? This PR is a bit more involved :)

**Comment by alextov** [comment]
@jseibert Sure.

Consider the following:

```swift
struct Book: Encodable {
    let title: String
    let author: String
    let year: String
}

let book = Book(title: "L’Étranger", author: "Albert Camus", year: "1942")

let encoder = XMLEncoder()
encoder.outputFormatting = [] // default value
print("outputFormatting = [] (default)")
print(String(data: try! encoder.encode(book, withRootKey: "book"), encoding: .utf8)!)
print()

encoder.outputFormatting = [.prettyPrinted]
print("outputFormatting = [.prettyPrinted]")
print(String(data: try! encoder.encode(book, withRootKey: "book"), encoding: .utf8)!)
print()

encoder.outputFormatting = [.sortedKeys]
print("outputFormatting = [.sortedKeys]")
print(String(data: try! encoder.encode(book, withRootKey: "book"), encoding: .utf8)!)
print()

encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
print("outputFormatting = [.prettyPrinted, .sortedKeys]")
print(String(data: try! encoder.encode(book, withRootKey: "book"), encoding: .utf8)!)
```

Output before this patch (`outputFormatting` has no effect, all combinations produce the same result):

```xml
outputFormatting = [] (default)
<book>
    <author>Albert Camus</author>
    <year>1942</year>
    <title>L’Étranger</title>
</book>

outputFormatting = [.prettyPrinted]
<book>
    <author>Albert Camus</author>
    <year>1942</year>
    <title>L’Étranger</title>
</book>

outputFormatting = [.sortedKeys]
<book>
    <author>Albert Camus</author>
    <year>1942</year>
    <title>L’Étranger</title>
</book>

outputFormatting = [.prettyPrinted, .sortedKeys]
<book>
    <author>Albert Camus</author>
    <year>1942</year>
    <title>L’Étranger</title>
</book>
```

Output after this patch:

```xml
outputFormatting = [] (default)
<book><author>Albert Camus</author><year>1942</year><title>L’Étranger</title></book>

outputFormatting = [.prettyPrinted]
<book>
    <author>Albert Camus</author>
    <year>1942</year>
    <title>L’Étranger</title>
</book>

outputFormatting = [.sortedKeys]
<book><author>Albert Camus</author><title>L’Étranger</title><year>1942</year></book>

outputFormatting = [.prettyPrinted, .sortedKeys]
<book>
    <author>Albert Camus</author>
    <title>L’Étranger</title>
    <year>1942</year>
</book>
```

Two notes regarding `.sortedKeys`: 

* My implementation uses standard `sort` method on the node dictionary which may or may not have an effect on performance when parsing large XML objects.
* Availability constraints for it have been clearly mirrored from those of standard `JSONEncoder.OutputFormatting`. I left it untouched, although I’m not convinced that it is applicable to `XMLEncoder`.

**Comment by alextov** [comment]
@jseibert What do you think?

**Comment by jseibert** [comment]
@alextov Apologies! I totally missed the notification of your earlier reply. Will review shortly.

File: Sources/XMLParsing/XMLStackParser.swift [hunk 1]
```diff
@@ -234,16 +237,30 @@ internal class _XMLElement {
...
             string += "</\(key)>"
         } else if !children.isEmpty {
-            string += ">\n"
+            string += ">"
+            if outputFormatting.contains(.prettyPrinted) {
+                string += "\n"
+            }
+
+            let childElements: [[_XMLElement]]
+            if #available(OSX 10.13, iOS 11.0, watchOS 4.0, tvOS 9.0, *), outputFormatting.contains(.sortedKeys) {
```
Comment (jseibert) [inline Sources/XMLParsing/XMLStackParser.swift]: Let's remove the #available checks. It will always be available, since you implemented them :)

**Comment by jseibert** [comment]
@alextov Overall this looks great! I'm happy to merge after you remove the availability checks? Those aren't needed here because you have implemented them, so they will never be unavailable. Can you remove those as well as the available declarations in the enum?

**Comment by alextov** [comment]
@jseibert I agree. I think these were mirrored from similar declarations in Swift’s own `JSONEncoder` class, which may use some internal APIs that were unavailable on older platforms. I was going to address it in my original pull request, but for some reason decided not to.

File: Sources/XMLParsing/XMLStackParser.swift [hunk 2]
```diff
@@ -234,16 +237,25 @@ internal class _XMLElement {
...
-            for childElement in children {
-                for child in childElement.value {
-                    string += child._toXMLString(indented: level + 1, withCDATA: cdata)
-                    string += "\n"
+            string += ">"
+            if outputFormatting.contains(.prettyPrinted) {
+                string += "\n"
+            }
+
+            let childElements = children.sorted { $0.key < $1.key } .map { $0.value }
```
Comment (jseibert) [inline Sources/XMLParsing/XMLStackParser.swift]: @alextov is this now always sorting the keys even if that option isn't set?

Comment (alextov) [reply Sources/XMLParsing/XMLStackParser.swift]: @jseibert Oh my, you’re right. Apologies! I have fixed the condition and pushed the changes.

**Comment by alextov** [comment]
@jseibert Sorry for that blunder. I fixed the issue, it should respect `outputFormatting` now.

**Pull request merged**